src/Sparrow/Allocators.Arena.cs

```diff
@@ -0,0 +1,336 @@
+        }
+
+        public void Reset(ref ArenaAllocator<TOptions> allocator)
+        {
+            // Is our strategy requiring us to change the size after having worked with it?
+            // The reasons to consider current buffer size and also the used amount of memory is very important to figure out
+            // what is the optimal policy for the arena size on the next round.
+            int preferredSize = allocator._options.GrowthStrategy.GetPreferredSize(allocator._currentBuffer.Size, allocator.InUse);
+
+            if (allocator._olderBuffers != null)
+            {
+                // Free old buffers not being used anymore. A single arena ought to be enough for anybody, you know ;) 
+                foreach (var unusedBuffer in allocator._olderBuffers)
+                {
+                    _used += unusedBuffer.Size;
+
+                    Pointer ptr = unusedBuffer;
+
+                    allocator.Allocated -= ptr.Size;
+                    allocator._nativeAllocator.Release(ref allocator._nativeAllocator, ref ptr);
+                }
+                allocator._olderBuffers.Clear();
+            }
+
+            // If we didnt use more memory than the currently allocated or we are in low memory mode
+            if (allocator.InUse <= allocator._currentBuffer.Size || Allocator.LowMemoryFlag.IsRaised())
+            {
+                // Override with the current setup
+                preferredSize = allocator._currentBuffer.Size;
+            }
+
+            if (allocator._currentBuffer.IsValid)
+            {
+                allocator.Allocated -= _currentBuffer.Size;
+                allocator._nativeAllocator.Release(ref allocator._nativeAllocator, ref _currentBuffer);                
+            }                
+
+            // We will always cap the size of the Arena toward the MaxArenaSize.
+            if (preferredSize > allocator._options.MaxArenaSize)
+                preferredSize = allocator._options.MaxArenaSize;
+
+            allocator._preferredSize = preferredSize;
+            allocator._currentBuffer = new Pointer();
+            allocator._ptrStart = allocator._ptrCurrent = null;
+            allocator._used = 0;
+
+            allocator.TotalAllocated = 0;
+            allocator.Allocated = 0;
+            allocator.InUse = 0;
+        }
+
+        public void Renew(ref ArenaAllocator<TOptions> allocator)
+        {
+            if (!allocator._currentBuffer.IsValid)
+            {
+                allocator._currentBuffer = allocator._nativeAllocator.Allocate(ref allocator._nativeAllocator, (int)allocator._preferredSize);
+                allocator.Allocated += allocator._currentBuffer.Size;
+            }
+
+            if (allocator._olderBuffers != null)
+            {
+                // Free old buffers not being used anymore. A single arena ought to be enough for anybody, you know ;) 
+                foreach (var unusedBuffer in allocator._olderBuffers)
```

Comment:
I changed the semantics of ``Reset`` and ``Renew`` to unify criteria. That was the second opinion I needed yesterday.

Reply:
What _are_ the semantics then, now? This need to be explicitly stated.

Reply:
The semantics are:
``Renew()`` ensures that if possible and supported by the allocator, no memory is going to be released to the underlying allocator. Let's say a pool, a pool unless is faced with memory pressure, it is not going to release the memory to the Arena Allocator or Native Allocator behind it if called ``Renew()`` but it will if called ``Reset``.

``Reset()`` is a fresh start, let's say that the ArenaAllocator has been running for a long time and now has an arena size (``.Allocated``) of 400Mb. With a Renew it will just move the pointer to zero and go on with its business, with Reset it will go, check the new preferred size given the behavior of the last round and return the memory to the underlying memory system if different. 
 
The current behavior of the allocators is confusing, because some behave like this, an others behave differently. What I am trying to do is to have a clear sematic for both operations. Now there is no reason why you would call ``Renew`` and ``Reset`` at the same time as we do in ``JsonContext``

Reply:
The problem we have here is that there are two distinct stages to deal with.
A context is checked out, used and returned.
The problem is what you are expected to do on the return.
You might want to return the memory back, retain it, etc. 

We need to play with the lifecycle of the context usage.

Given `AllocateContext` and `ReturnContext` in their use in web requests, for example, what will be called on either?

Reply:
The idea behind a clear cut definition of what each operation do is to avoid the current situation, where each allocator decides how it interprete each one. It is the job of ``AllocateContext`` and ``ReturnContext`` to assess which one they should call based on usage patterns.